gitcache/main.go

```diff
@@ -0,0 +1,149 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/mholt/archiver"
+)
+
+func main() {
+	blob := os.Getenv("BLOB_URL")
+
+	if blob == "" {
+		log.Panic("BLOB_URL env is not set.")
+	}
+
+	repo := &RepoURL{}
+	err := repo.Set(os.Args[1])
+	if err != nil {
+		log.Panic(err)
+	}
+
+	path, err := os.Getwd()
+	if err != nil {
+		log.Panic(err)
+	}
+
+	dir, err := ioutil.TempDir(path, repo.Owner+repo.Owner)
+	if err != nil {
+		log.Panic(err)
+	}
+
+	tarDir, err := ioutil.TempDir(path, repo.Owner+repo.Owner+"tar")
+	if err != nil {
+		log.Panic(err)
+	}
+
+	defer os.RemoveAll(dir)
+	defer os.RemoveAll(tarDir)
+
+	// Clone the given repository to the given directory
+	r, err := git.PlainClone(dir, false, &git.CloneOptions{
+		URL:   fmt.Sprintf("http://%s/%s/%s", repo.Host, repo.Owner, repo.Repo),
+		Depth: 1, // Just fetch the last commit
+	})
+	if err != nil {
+		log.Panic(err)
+	}
+	ref, err := r.Head()
+	if err != nil {
+		log.Panic(err)
+	}
+
+	// ... retrieving the commit object
+	commit, err := r.CommitObject(ref.Hash())
+	if err != nil {
+		log.Panic(err)
+	}
+
+	// opening the blob
+	bucket, e := New(blob)
+
+	if e != nil {
+		log.Panic(e)
+	}
+
+	tpath := path.Join(tarDir, fmt.Sprintf("%s.tar.gz", repo.Repo))
+	// storing the last commit to the blob
+	err = bucket.Set(fmt.Sprintf("gitcache/%s/%s/lastcommit", repo.Owner, repo.Repo), []byte(fmt.Sprintf("%64b\n", commit.Author.When.Unix())))
+	if err != nil {
+		log.Panic(err)
+	}
+
+	now := time.Now().Unix()
+	err = bucket.Set(fmt.Sprintf("gitcache/%s/%s/lastsync", repo.Owner, repo.Repo), []byte(fmt.Sprintf("%64b\n", now)))
+	if err != nil {
+		log.Panic(err)
+	}
+
+	// removing the .git folder as it is not required for the tar ball
+	err = os.RemoveAll(path.Join(dir, ".git"))
+	if err != nil {
+		log.Panic(err)
+	}
+
+	// creating an archive
+	t := archiver.NewTarGz()
+
+	err = t.Archive([]string{dir}, tpath)
+
+	if err != nil {
+		log.Panic(err)
+	}
+
+	data, err := ioutil.ReadFile(tpath)
+	if err != nil {
+		log.Panic(err)
+	}
+
+	// storing the archive to the blob
+	err = bucket.Set(fmt.Sprintf("gitcache/%s/%s/tar", repo.Owner, repo.Repo), data)
+	if err != nil {
+		log.Panic(err)
+	}
+}
+
+type RepoURL struct {
```

Comment:
Not reusing the `RepoURL` struct . https://go-proverbs.github.io/ `A little copying is better than a little dependency.`

We could probably pull out `gitcache` into a separate `repo`